services/system-user-service/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/CognitoUserDataAccessLayer.java

```diff
@@ -0,0 +1,291 @@
+import java.time.ZoneOffset;
+import java.util.*;
+
+public class CognitoUserDataAccessLayer implements SystemUserDataAccessLayer {
+
+    private final static Logger LOGGER = LoggerFactory.getLogger(CognitoUserDataAccessLayer.class);
+    private final static String COGNITO_USERPOOL_ID = System.getenv("SAAS_BOOST_USERPOOL_ID");
+    private final CognitoIdentityProviderClient cognito;
+
+    public CognitoUserDataAccessLayer() {
+        if (Utils.isBlank(COGNITO_USERPOOL_ID)) {
+            throw new IllegalStateException("Missing required environment variable SAAS_BOOST_USERPOOL_ID");
+        }
+        this.cognito = Utils.sdkClient(CognitoIdentityProviderClient.builder(),
+                CognitoIdentityProviderClient.SERVICE_NAME);
+    }
+
+    public List<SystemUser> getUsers() {
+        LOGGER.info("UserServiceDAL::getUsers");
+        List<SystemUser> users = new ArrayList<>();
+        String nextToken = null;
+        do {
+            try {
+                ListUsersResponse response = cognito.listUsers(ListUsersRequest.builder()
+                        .userPoolId(COGNITO_USERPOOL_ID)
+                        .paginationToken(nextToken)
+                        .build()
+                );
+                nextToken = response.paginationToken();
+                for (UserType userType : response.users()) {
+                    users.add(fromUserType(userType));
+                }
+            } catch (SdkServiceException cognitoError) {
+                LOGGER.error("cognito-idp:ListUsers", cognitoError);
+                LOGGER.error(Utils.getFullStackTrace(cognitoError));
+                throw cognitoError;
+            }
+        } while (Utils.isNotEmpty(nextToken));
+        return users;
+    }
+
+    public SystemUser getUser(String username) {
+        LOGGER.info("UserServiceDAL::getUser");
+        SystemUser user = null;
+        try {
+            AdminGetUserResponse response = cognito.adminGetUser(request -> request
+                    .username(username)
+                    .userPoolId(COGNITO_USERPOOL_ID)
+            );
+            // Cognito's SDK was clearly not written to be user friendly...
+            UserType userType = UserType.builder()
+                    .userCreateDate(response.userCreateDate())
+                    .userLastModifiedDate(response.userLastModifiedDate())
+                    .enabled(response.enabled())
+                    .attributes(response.userAttributes())
+                    .userStatus(response.userStatus())
+                    .userStatus(response.userStatusAsString())
+                    .username(response.username())
+                    .mfaOptions(response.mfaOptions())
+                    .build();
+            user = fromUserType(userType);
+        } catch (SdkServiceException cognitoError) {
+            LOGGER.error("cognito-idp:AdminGetUser", cognitoError);
+            LOGGER.error(Utils.getFullStackTrace(cognitoError));
+            //throw cognitoError;
```

Comment:
why not throw the exception here?

Reply:
This is old stuff. I'd guess it's because the service calling this DAL is returning a 404 if we can't load a user from the User Pool by username instead of a 5xx. We can dig into what kinds of exceptions are thrown by Cognito and think about what kind of experience that should mean for the client of the API.